msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs

```diff
@@ -0,0 +1,153 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.Features.DeviceCode;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Internal.Requests;
+using Microsoft.Identity.Core;
+using Microsoft.Identity.Core.Helpers;
+using Microsoft.Identity.Core.Http;
+using Microsoft.Identity.Core.Instance;
+using Microsoft.Identity.Core.OAuth2;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Test.Microsoft.Identity.Core.Unit.Mocks;
+
+namespace Test.MSAL.NET.Unit.RequestsTests
+{
+    [TestClass]
+    public class DeviceCodeRequestTests
+    {
+        private TokenCache _cache;
+        private readonly MyReceiver _myReceiver = new MyReceiver();
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            RequestTestsCommon.InitializeRequestTests();
+            Telemetry.GetInstance().RegisterReceiver(_myReceiver.OnEvents);
+
+            _cache = new TokenCache();
+        }
+
+        [TestCleanup]
+        public void TestCleanup()
+        {
+            _cache.tokenCacheAccessor.AccessTokenCacheDictionary.Clear();
+            _cache.tokenCacheAccessor.RefreshTokenCacheDictionary.Clear();
+        }
+
+        private HttpResponseMessage CreateDeviceCodeResponseSuccessMessage()
+        {
+            return MockHelpers.CreateSuccessResponseMessage(
+                "{\"user_code\":\"B6SUYU5PL\",\"device_code\":\"BAQABAAEAAADXzZ3ifr-GRbDT45zNSEFEfU4P-bZYS1vkvv8xiXdb1_zX2xAcdcfEoei1o-t9-zTB9sWyTcddFEWahP1FJJJ_YVA1zvPM2sV56d_8O5G23ti5uu0nRbIsniczabYYEr-2ZsbgRO62oZjKlB1zF3EkuORg2QhMOjtsk-KP0aw8_iAA\",\"verification_url\":\"https://microsoft.com/devicelogin\",\"expires_in\":\"900\",\"interval\":\"5\",\"message\":\"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code B6SUYU5PL to authenticate.\"}");
+        }
+
+        [TestMethod]
+        [TestCategory("DeviceCodeRequestTests")]
+        public void TestDeviceCodeAuthSuccess()
+        {
+            Authority authority = Authority.CreateAuthority(TestConstants.AuthorityHomeTenant, false);
+            _cache = new TokenCache()
+            {
+                ClientId = TestConstants.ClientId
+            };
+
+            AuthenticationRequestParameters parameters = new AuthenticationRequestParameters()
+            {
+                Authority = authority,
+                // todo: what is this? SliceParameters = "key1=value1%20with%20encoded%20space&key2=value2",
+                ClientId = TestConstants.ClientId,
+                Scope = TestConstants.Scope,
+                TokenCache = _cache,
```

Comment:
assert token case before and after state? I suppose we are not caching tokens for Device Code?

Reply:
We are writing them to the cache, but we don't try to read them from the cache.

Reply:
Kind of makes sense to enable using the RT to for silent calls going forward. Seems like a good scenario to validate in the e2e testing. Device Code would be using the application cache given there is no user involved. If we can't make silent calls based on client_id alone then I suppose token cashing isn't useful.

---
In reply to: [220764186](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1275#discussion_r220764186) [](ancestors = 220764186)